django/db/backends/mysql/client.py

```diff
@@ -45,8 +54,10 @@ def settings_to_cmd_args(cls, settings_dict, parameters):
         if db:
             args += [db]
         args.extend(parameters)
-        return args
+        return args, env
 
     def runshell(self, parameters):
-        args = DatabaseClient.settings_to_cmd_args(self.connection.settings_dict, parameters)
-        subprocess.run(args, check=True)
+        args, env = DatabaseClient.settings_to_cmd_args_env(self.connection.settings_dict, parameters)
```

Comment:
I wonder if you have considered catching the `CalledProcessError`, modifying the message, and re-throwing it (or throwing a different `CalledProcessError` instance)?

It has the advantage of avoiding the `MYSQL_PWD` env var that MySQL advises against (and even [calls "insecure"](https://dev.mysql.com/doc/refman/8.0/en/environment-variables.html)), and means we don't need to make changes to the signatures of these DB backends.

I must admit, I'm not sure if it would be much better, as it runs the risk of the same issue occurring for other errors that may be raised here. Modifying the error as it's raised could be considered a little distasteful, too, I suppose.

Reply:
@meshy thanks for taking a look at the patch. This re-reraise approach was mentioned during the initial exchange with the security team but it didn't get feedback about it.

It also didn't address the issue mentioned in the original ticket report which is that some OS don't obfuscate command line arguments which means it might still be visible for other processes (basically [#24999](https://code.djangoproject.com/ticket/24999)).

To quote @apollo13 for the security@djangoproject.com exchange

> From a security perspective "file > envvar > cmd arg", so using env vars over cmd args is __always__ an improvement.

Now the ideal solution in this case would likely be to use a file to store the credentials but it seems the arguments brought on MySQL documentation against using the `MYSQL_PWD` environment variable are also applicable to using the `--password` command line argument.

Moreover it seems the environment variable leak issue is only present on old UNIX system such as [AIX](https://unix.stackexchange.com/questions/369566/why-is-passing-the-secrets-via-environmental-variables-considered-extremely-ins).

Maybe we just bite the bullet and switch all backends to use configuration files like it uses to be case for PostgreSQL before https://github.com/django/django/commit/cf826c9a91015c8da2ad4910b12e2ed83e2fb20f#diff-f1608df8f66bd1290471472cf2ef11b5. It'd be great if we could be more coherent with our we treat this security issue; using `PGPASSWORD` is not more secure than using `MYSQL_PWD`.

Reply:
FWIW, mysql "masks" the password on the command line by replacing it with `X` for every character -- at least on system where it is possible. Since a cmd line argument usually leaks on every system I'd consider cmd line arguments highly insecure (no matter what). But even if the replacing of the password works properly on every system the error python raises would still contain it.

An environment variable is usually somewhat secure (ie other users are generally not able to view them). A file would even fix those remaining issues but can be hard to implement -- one should just use a (non-swapping) ram-fs backed tempfile, otherwise you would persist the password on disk where it might be recoverable. But I'd argue that the last issue is not really such an issue because the password has to be on the disk already (settings file ;)) or somehow passed into it (and even if it comes as env variable into the settings file it usually is coming from yet another file on this PC).

All in all I do not think that dbshell is used that much on production systems (or at least should not be), but making sure that we have no obvious leaks like in exceptions is worth it.

Reply:
Thanks for explaining the thought process. That all sounds reasonable to me.

I don't have any more questions!